File: frontend/src/types/trip.ts
/**
 * Trip TypeScript Type Definitions
 *
 * Type definitions for the Travel Diary feature.
 * Based on backend API contracts from Feature 002 (Travel Diary Backend).
 */

// ============================================================================
// Core Entities (API Responses)
// ============================================================================

/**
 * Complete trip data from backend API
 *
 * Used in:
 * - TripDetailPage (view full trip)
 * - TripEditPage (edit existing trip)
 * - After creating/updating a trip
 */
export interface Trip {
  /** Unique trip identifier (UUID) */
  trip_id: string;

  /** Trip owner's user ID (UUID) */
  user_id: string;

  /** Trip title (1-200 characters) */
  title: string;

  /** Trip description (HTML, sanitized by backend, max 50000 chars) */
  description: string;

  /** Publication status */
  status: 'draft' | 'published';

  /** Trip start date (ISO 8601: YYYY-MM-DD) */
  start_date: string;

  /** Trip end date (ISO 8601: YYYY-MM-DD, null if single-day trip) */
  end_date: string | null;

  /** Distance in kilometers (0.1-10000, null if not provided) */
  distance_km: number | null;

  /** Difficulty level (null if not provided) */
  difficulty: TripDifficulty | null;

  /** Trip creation timestamp (ISO 8601: YYYY-MM-DDTHH:mm:ssZ) */
  created_at: string;

  /** Last update timestamp (ISO 8601: YYYY-MM-DDTHH:mm:ssZ) */
  updated_at: string;

  /** Publication timestamp (ISO 8601, null if draft) */
  published_at: string | null;

  /** Trip photos (ordered by display_order) */
  photos: TripPhoto[];

  /** Trip locations/waypoints (ordered by sequence) */
  locations: TripLocation[];

  /** Trip tags for categorization */
  tags: Tag[];
}

/**
 * Trip summary for list/grid views
 *
 * Used in:
 * - TripsListPage (browse all trips)
 * - ProfilePage (user's recent trips)
 * - Search results
 */
export interface TripListItem {
  /** Unique trip identifier (UUID) */
  trip_id: string;

  /** Trip owner's user ID (UUID) */
  user_id: string;

  /** Trip title */
  title: string;

  /** Trip start date (ISO 8601: YYYY-MM-DD) */
  start_date: string;

  /** Distance in kilometers (null if not provided) */
  distance_km: number | null;

  /** Publication status */
  status: 'draft' | 'published';

  /** Number of photos attached to trip */
  photo_count: number;

  /** Tag names only (for filtering, lighter payload) */
  tag_names: string[];

  /** First photo thumbnail URL (null if no photos) */
  thumbnail_url: string | null;

  /** Trip creation timestamp (ISO 8601: YYYY-MM-DDTHH:mm:ssZ) */
  created_at: string;
}

/**
 * Legacy alias for TripListItem (backward compatibility)
 * @deprecated Use TripListItem instead
 */
export type TripSummary = TripListItem;

/**
 * Trip photo data
 *
 * Used in:
 * - TripGallery (photo grid with lightbox)
 * - PhotoUploader (uploaded photo preview)
 */
export interface TripPhoto {
  /** Unique photo identifier (UUID) */
  photo_id: string;

  /** URL to optimized photo (max 2000px width) */
  photo_url: string;

  /** URL to thumbnail (400x400px) */
  thumbnail_url: string;

  /** Optional photo caption (max 500 chars) */
  caption: string | null;

  /** Display order in gallery (0-based, reorderable) */
  display_order: number;

  /** Optimized photo width in pixels */
  width: number;

  /** Optimized photo height in pixels */
  height: number;
}

/**
 * Trip tag for categorization
 *
 * Used in:
 * - TagInput (autocomplete suggestions)
 * - TripFilters (clickable tag chips)
 * - Trip detail view (tag chips)
 */
export interface Tag {
  /** Unique tag identifier (UUID) */
  tag_id: string;

  /** Display name (preserves original casing, e.g., "Vías Verdes") */
  name: string;

  /** Normalized name for matching (lowercase, e.g., "vias verdes") */
  normalized: string;

  /** Number of trips using this tag (for popularity sorting) */
  usage_count: number;
}

/**
 * Trip location/waypoint
 *
 * Used in:
 * - TripMap (display markers on map)
 * - Location list in trip details
 */
export interface TripLocation {
  /** Unique location identifier (UUID) */
  location_id: string;

  /** Location name (e.g., "Baeza", "Camino de Santiago") */
  name: string;

  /** Latitude coordinate (decimal degrees, null if not geocoded) */
  latitude: number | null;

  /** Longitude coordinate (decimal degrees, null if not geocoded) */
  longitude: number | null;

  /** Order in route (0 = start, 1 = next waypoint, etc.) */
  sequence: number;
}

// ============================================================================
// Request/Input Schemas
// ============================================================================

/**
 * Input data for creating a new trip
 *
 * Used in:
 * - TripFormWizard (multi-step form submission)
 * - Step 4 (Review & Publish)
 */
export interface TripCreateInput {
  /** Trip title (1-200 characters, required) */
  title: string;

  /** Trip description (1-50000 characters, required, HTML allowed) */
  description: string;

  /** Trip start date (ISO 8601: YYYY-MM-DD, cannot be future) */
  start_date: string;

  /** Trip end date (ISO 8601: YYYY-MM-DD, must be >= start_date) */
  end_date: string | null;

  /** Distance in kilometers (0.1-10000, optional) */
  distance_km: number | null;

  /** Difficulty level (optional) */
  difficulty: TripDifficulty | null;

  /** Locations visited during trip (max 50) */
  locations: LocationInput[];

  /** Tags for categorization (max 10 tags, max 50 chars each) */
  tags: string[];
}

/**
 * Input data for updating an existing trip
 *
 * All fields are optional - only provided fields will be updated.
 *
 * Used in:
 * - TripFormWizard (edit mode)
 * - Step 4 (Review & Publish)
 */
export interface TripUpdateInput {
  /** Trip title (1-200 characters) */
  title?: string;

  /** Trip description (1-50000 characters, HTML allowed) */
  description?: string;

  /** Trip start date (ISO 8601: YYYY-MM-DD) */
  start_date?: string;

  /** Trip end date (ISO 8601: YYYY-MM-DD, must be >= start_date) */
  end_date?: string | null;

  /** Distance in kilometers (0.1-10000) */
  distance_km?: number | null;

  /** Difficulty level */
  difficulty?: TripDifficulty | null;

  /** Locations (replaces existing) */
  locations?: LocationInput[];

  /** Tags (replaces existing, max 10) */
  tags?: string[];

  /** Timestamp when client loaded the trip (optimistic locking) */
  client_updated_at?: string;
}

/**
 * Location input for trip forms
 *
 * Used in:
 * - Step 1 (Basic Info) - location input field
 * - Trip creation/update requests
 */
export interface LocationInput {
  /** Location name (1-200 characters, required) */
  name: string;

  /** Country name (max 100 characters, optional) */
  country?: string;

  /** Latitude in decimal degrees (-90 to 90, optional) */
  latitude?: number | null;

  /** Longitude in decimal degrees (-180 to 180, optional) */
  longitude?: number | null;
}

// ============================================================================
// Response Schemas
// ============================================================================

/**
 * Paginated trip list response
 *
 * Used in:
 * - TripsListPage (browse trips with filters)
 * - useTripList hook (pagination logic)
 */
export interface TripListResponse {
  /** List of trip summaries for current page */
  trips: TripListItem[];

  /** Total number of trips matching filter (for pagination UI) */
  total: number;

  /** Page size used for this request */
  limit: number;

  /** Pagination offset used for this request */
  offset: number;
}

/**
 * Standard API error response
 *
 * Used in:
 * - All API service error handlers
 * - Error toast notifications
 */
export interface ApiError {
  /** Always false for errors */
  success: false;

  /** Null data payload */
  data: null;

  /** Error details */
  error: {
    /** Error code (e.g., "VALIDATION_ERROR", "NOT_FOUND") */
    code: string;

    /** User-friendly error message in Spanish */
    message: string;

    /** Optional field name for validation errors */
    field?: string;
  };
}

// ============================================================================
// Form State (Client-Only)
// ============================================================================

/**
 * Form state for TripFormWizard (all 4 steps combined)
 *
 * Used in:
 * - TripFormWizard (parent component state)
 * - All step components (Step1-4)
 * - Zod validation schema
 */
export interface TripFormData {
  // Step 1: Basic Info
  title: string;
  start_date: string; // HTML date input format: YYYY-MM-DD
  end_date: string;
  distance_km: string; // String in form, parsed to number on submit
  difficulty: TripDifficulty | '';

  // Step 2: Story & Tags
  description: string;
  tags: string[]; // Array of tag names

  // Step 3: Photos
  // Photos handled separately via PhotoUploader (file uploads)

  // Step 4: Review
  // Read-only summary of all fields above
}

/**
 * Photo upload state for PhotoUploader component
 *
 * Used in:
 * - PhotoUploader (Step 3)
 * - useTripPhotos hook
 */
export interface PhotoUploadState {
  /** Selected files for upload */
  files: File[];

  /** Upload progress per file (0-100) */
  progress: Record<string, number>; // key: file.name, value: percentage

  /** Upload status per file */
  status: Record<string, 'pending' | 'uploading' | 'success' | 'error'>;

  /** Uploaded photo responses from backend */
  uploadedPhotos: TripPhoto[];

  /** Error messages per file */
  errors: Record<string, string>;
}

// ============================================================================
// Enums and Constants
// ============================================================================

/**
 * Trip publication status
 */
export type TripStatus = 'draft' | 'published';

/**
 * Trip difficulty levels
 */
export type TripDifficulty = 'easy' | 'moderate' | 'difficult' | 'very_difficult';

/**
 * Difficulty labels in Spanish
 */
export const DIFFICULTY_LABELS: Record<TripDifficulty, string> = {
  easy: 'Fácil',
  moderate: 'Moderada',
  difficult: 'Difícil',
  very_difficult: 'Muy Difícil',
};

/**
 * Difficulty CSS class names
 */
export const DIFFICULTY_CLASSES: Record<TripDifficulty, string> = {
  easy: 'difficulty-badge--easy',
  moderate: 'difficulty-badge--moderate',
  difficult: 'difficulty-badge--difficult',
  very_difficult: 'difficulty-badge--very-difficult',
};

// ============================================================================
// Public Feed Types (Feature 013 - Public Trips Feed)
// ============================================================================

/**
 * User summary for public trip feed
 *
 * Minimal user info displayed on public trip cards - only public profile data.
 *
 * Used in:
 * - PublicFeedPage (homepage)
 * - PublicTripCard component
 */
export interface PublicUserSummary {
  /** Unique user identifier (UUID) */
  user_id: string;

  /** Username (for profile link) */
  username: string;

  /** Profile photo URL (null if no photo) */
  profile_photo_url: string | null;
}

/**
 * Photo summary for public trip feed
 *
 * Minimal photo info for trip cards - only first photo thumbnail.
 *
 * Used in:
 * - PublicTripCard (display first photo)
 */
export interface PublicPhotoSummary {
  /** URL to optimized photo */
  photo_url: string;

  /** URL to thumbnail */
  thumbnail_url: string;
}

/**
 * Location summary for public trip feed
 *
 * Minimal location info for trip cards - only first location name.
 *
 * Used in:
 * - PublicTripCard (display first location)
 */
export interface PublicLocationSummary {
  /** Location name (e.g., "Baeza, España") */
  name: string;
}

/**
 * Trip summary for public feed (Feature 013)
 *
 * Optimized for homepage display - shows only essential trip data.
 *
 * Requirements (FR-002):
 * - título, foto, distancia, location (primera), fecha, autor
 *
 * Used in:
 * - PublicFeedPage (homepage trip grid)
 * - PublicTripCard component
 */
export interface PublicTripSummary {
  /** Unique trip identifier (UUID) */
  trip_id: string;

  /** Trip title */
  title: string;

  /** Trip start date (ISO 8601: YYYY-MM-DD) */
  start_date: string;

  /** Distance in kilometers (null if not provided) */
  distance_km: number | null;

  /** First photo thumbnail (null if no photos) */
  photo: PublicPhotoSummary | null;

  /** First location (null if no locations) */
  location: PublicLocationSummary | null;

  /** Trip author summary */
  author: PublicUserSummary;

  /** Publication timestamp (ISO 8601, for sorting) */
  published_at: string;
}

/**
 * Pagination metadata for public feed
 *
 * Provides information for client-side pagination controls.
 *
 * Used in:
 * - PublicFeedPage (pagination UI)
 * - usePublicTrips hook
 */
export interface PaginationInfo {
  /** Total number of trips matching filter */
  total: number;

  /** Current page number (1-indexed) */
  page: number;

  /** Page size (trips per page, max 50) */
  limit: number;

  /** Total number of pages */
  total_pages: number;
}

/**
 * Paginated response for public trips feed (Feature 013)
 *
 * Main response schema for GET /trips/public endpoint.
 *
 * Used in:
 * - PublicFeedPage (homepage)
 * - usePublicTrips hook
 */
export interface PublicTripListResponse {
  /** List of public trip summaries */
  trips: PublicTripSummary[];

  /** Pagination metadata */
  pagination: PaginationInfo;
}

File: backend/src/models/trip.py
"""
SQLAlchemy models for Travel Diary feature.

Defines Trip, TripPhoto, Tag, TripTag, and TripLocation models.
Supports both PostgreSQL (production) and SQLite (development/testing).
"""

import enum
import uuid
from datetime import UTC, datetime

from sqlalchemy import (
    Column,
    Date,
    DateTime,
    Enum,
    Float,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
)
from sqlalchemy.orm import Mapped, relationship

from src.database import Base


def generate_uuid() -> str:
    """Generate UUID string for primary keys."""
    return str(uuid.uuid4())


class TripStatus(str, enum.Enum):
    """Trip publication status."""

    DRAFT = "draft"  # Not published, only visible to owner
    PUBLISHED = "published"  # Published, visible to others


class TripDifficulty(str, enum.Enum):
    """Trip difficulty levels."""

    EASY = "easy"  # Fácil
    MODERATE = "moderate"  # Moderada
    DIFFICULT = "difficult"  # Difícil
    VERY_DIFFICULT = "very_difficult"  # Muy Difícil


class Trip(Base):
    """
    Trip model - Main entity for travel diary entries.

    Represents a cycling trip with all its details: route, photos, tags, locations.
    """

    __tablename__ = "trips"

    # Primary key
    trip_id = Column(String(36), primary_key=True, default=generate_uuid)

    # Foreign keys
    user_id = Column(String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)

    # Core fields
    title = Column(String(200), nullable=False)  # Trip title
    description = Column(Text, nullable=False)  # Rich text description (HTML)
    status = Column(
        Enum(TripStatus, native_enum=False, length=20),
        nullable=False,
        default=TripStatus.DRAFT,
    )

    # Trip details
    start_date = Column(Date, nullable=False)  # When trip started
    end_date = Column(Date, nullable=True)  # When trip ended (optional)
    distance_km = Column(Float, nullable=True)  # Distance in kilometers
    difficulty = Column(
        Enum(TripDifficulty, native_enum=False, length=20),
        nullable=True,
    )

    # Metadata
    created_at = Column(DateTime(timezone=True), nullable=False, default=lambda: datetime.now(UTC))
    updated_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(UTC),
        onupdate=lambda: datetime.now(UTC),
    )
    published_at = Column(DateTime(timezone=True), nullable=True)  # When first published

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="trips")  # type: ignore
    photos: Mapped[list["TripPhoto"]] = relationship(
        "TripPhoto",
        back_populates="trip",
        cascade="all, delete-orphan",
        order_by="TripPhoto.order",
    )
    locations: Mapped[list["TripLocation"]] = relationship(
        "TripLocation",
        back_populates="trip",
        cascade="all, delete-orphan",
        order_by="TripLocation.sequence",
    )
    trip_tags: Mapped[list["TripTag"]] = relationship(
        "TripTag",
        back_populates="trip",
        cascade="all, delete-orphan",
    )

    # Social network relationships (Feature 004)
    likes: Mapped[list["Like"]] = relationship(
        "Like",
        back_populates="trip",
        cascade="all, delete-orphan",
    )

    comments: Mapped[list["Comment"]] = relationship(
        "Comment",
        back_populates="trip",
        cascade="all, delete-orphan",
    )

    shares: Mapped[list["Share"]] = relationship(
        "Share",
        back_populates="trip",
        cascade="all, delete-orphan",
    )

    # Indexes
    __table_args__ = (
        Index("idx_trip_user_status", "user_id", "status"),  # List user's trips by status
        Index("idx_trip_created", "created_at"),  # Sort by creation date
        Index("idx_trip_published", "published_at"),  # Filter published trips
    )

    def __repr__(self) -> str:
        return f"<Trip(trip_id={self.trip_id}, title={self.title}, status={self.status})>"


class TripPhoto(Base):
    """
    TripPhoto model - Photos attached to trips.

    Stores both optimized and thumbnail versions of trip photos.
    """

    __tablename__ = "trip_photos"

    # Primary key
    photo_id = Column(String(36), primary_key=True, default=generate_uuid)

    # Foreign keys
    trip_id = Column(String(36), ForeignKey("trips.trip_id", ondelete="CASCADE"), nullable=False)

    # Photo storage paths
    photo_url = Column(String(500), nullable=False)  # Path to optimized version
    thumb_url = Column(String(500), nullable=False)  # Path to thumbnail

    # Photo metadata
    caption = Column(String(500), nullable=True)  # Optional caption
    order = Column(
        Integer, nullable=False, default=0
    )  # Order in gallery (renamed from display_order)
    file_size = Column(Integer, nullable=False, default=0)  # File size in bytes
    width = Column(Integer, nullable=False, default=0)  # Image width in pixels
    height = Column(Integer, nullable=False, default=0)  # Image height in pixels

    # Metadata
    uploaded_at = Column(DateTime(timezone=True), nullable=False, default=lambda: datetime.now(UTC))

    # Relationships
    trip: Mapped["Trip"] = relationship("Trip", back_populates="photos")

    # Indexes
    __table_args__ = (
        Index("idx_photo_trip", "trip_id"),  # Find photos by trip
        Index("idx_photo_order", "trip_id", "order"),  # Ordered gallery
    )

    # Property mappings for schema compatibility
    @property
    def id(self) -> str:
        """Map photo_id to id for backward compatibility."""
        return self.photo_id

    @property
    def thumbnail_url(self) -> str:
        """Map thumb_url to thumbnail_url for schema compatibility."""
        return self.thumb_url

    @thumbnail_url.setter
    def thumbnail_url(self, value: str) -> None:
        """Allow setting thumbnail_url by writing to thumb_url."""
        self.thumb_url = value

    @property
    def display_order(self) -> int:
        """Map order to display_order for schema compatibility."""
        return self.order

    @display_order.setter
    def display_order(self, value: int) -> None:
        """Allow setting display_order by writing to order."""
        self.order = value

    def __repr__(self) -> str:
        return f"<TripPhoto(photo_id={self.photo_id}, trip_id={self.trip_id})>"


class Tag(Base):
    """
    Tag model - Tags for categorizing trips.

    Uses normalized lowercase for matching while preserving display format.
    """

    __tablename__ = "tags"

    # Primary key
    tag_id = Column(String(36), primary_key=True, default=generate_uuid)

    # Tag names
    name = Column(String(50), nullable=False, unique=True)  # Display name (original case)
    normalized = Column(String(50), nullable=False, unique=True)  # Lowercase for matching

    # Usage counter
    usage_count = Column(Integer, nullable=False, default=0)  # How many trips use this tag

    # Metadata
    created_at = Column(DateTime(timezone=True), nullable=False, default=lambda: datetime.now(UTC))

    # Relationships
    trip_tags: Mapped[list["TripTag"]] = relationship(
        "TripTag",
        back_populates="tag",
        cascade="all, delete-orphan",
    )

    # Indexes
    __table_args__ = (
        Index("idx_tag_normalized", "normalized"),  # Fast case-insensitive lookup
        Index("idx_tag_usage", "usage_count"),  # Popular tags
    )

    def __repr__(self) -> str:
        return f"<Tag(tag_id={self.tag_id}, name={self.name}, usage={self.usage_count})>"


class TripTag(Base):
    """
    TripTag model - Many-to-many association between Trips and Tags.

    Junction table linking trips to their tags.
    """

    __tablename__ = "trip_tags"

    # Composite primary key
    trip_id = Column(String(36), ForeignKey("trips.trip_id", ondelete="CASCADE"), primary_key=True)
    tag_id = Column(String(36), ForeignKey("tags.tag_id", ondelete="CASCADE"), primary_key=True)

    # Metadata
    created_at = Column(DateTime(timezone=True), nullable=False, default=lambda: datetime.now(UTC))

    # Relationships
    trip: Mapped["Trip"] = relationship("Trip", back_populates="trip_tags")
    tag: Mapped["Tag"] = relationship("Tag", back_populates="trip_tags")

    # Indexes
    __table_args__ = (
        Index("idx_trip_tag_trip", "trip_id"),  # Find tags for a trip
        Index("idx_trip_tag_tag", "tag_id"),  # Find trips with a tag
    )

    def __repr__(self) -> str:
        return f"<TripTag(trip_id={self.trip_id}, tag_id={self.tag_id})>"


class TripLocation(Base):
    """
    TripLocation model - Locations/waypoints along the trip route.

    Stores location name and optional coordinates (if geocoded).
    """

    __tablename__ = "trip_locations"

    # Primary key
    location_id = Column(String(36), primary_key=True, default=generate_uuid)

    # Foreign keys
    trip_id = Column(String(36), ForeignKey("trips.trip_id", ondelete="CASCADE"), nullable=False)

    # Location data
    name = Column(
        String(200), nullable=False
    )  # Location name (e.g., "Madrid", "Camino de Santiago")
    latitude = Column(Float, nullable=True)  # Decimal degrees (optional, from geocoding)
    longitude = Column(Float, nullable=True)  # Decimal degrees (optional, from geocoding)
    sequence = Column(Integer, nullable=False)  # Order along route (0 = start, 1 = next, etc.)

    # Metadata
    created_at = Column(DateTime(timezone=True), nullable=False, default=lambda: datetime.now(UTC))

    # Relationships
    trip: Mapped["Trip"] = relationship("Trip", back_populates="locations")

    # Indexes
    __table_args__ = (
        Index("idx_location_trip", "trip_id"),  # Find locations for a trip
        Index("idx_location_sequence", "trip_id", "sequence"),  # Ordered route
    )

    def __repr__(self) -> str:
        return (
            f"<TripLocation(location_id={self.location_id}, name={self.name}, seq={self.sequence})>"
        )
